server/build_event_protocol/build_event_handler/build_event_handler_test.go

```diff
@@ -1214,6 +1292,158 @@ func TestRetryOnOldDisconnect(t *testing.T) {
+						Metadata: map[string]string{"ROLE": "CI"},
+					}},
+				},
+				&bspb.BuildEvent{
+					Id: &bspb.BuildEventId{Id: &bspb.BuildEventId_WorkspaceStatus{}},
+					Payload: &bspb.BuildEvent_WorkspaceStatus{WorkspaceStatus: &bspb.WorkspaceStatus{
+						Item: []*bspb.WorkspaceStatus_Item{
+							{Key: "REPO_URL", Value: "https://github.com/testowner/testrepo.git"},
+							{Key: "COMMIT_SHA", Value: "0c894fe31c2e91d59cb1a59bb25aaa78089919c2"},
+						},
+					}},
+				},
+			},
+		},
+		// TODO: this test fails - fix
+		// {
+		// 	name: "WorkspaceStatusThenBuildMetadata",
+		// 	metadataEvents: []*bspb.BuildEvent{
+		// 		&bspb.BuildEvent{
+		// 			Id: &bspb.BuildEventId{Id: &bspb.BuildEventId_Pattern{Pattern: &bspb.BuildEventId_PatternExpandedId{
+		// 				Pattern: []string{"//..."},
+		// 			}}},
+		// 		},
+		// 		&bspb.BuildEvent{
+		// 			Id: &bspb.BuildEventId{Id: &bspb.BuildEventId_WorkspaceStatus{}},
+		// 			Payload: &bspb.BuildEvent_WorkspaceStatus{WorkspaceStatus: &bspb.WorkspaceStatus{
+		// 				Item: []*bspb.WorkspaceStatus_Item{
+		// 					{Key: "REPO_URL", Value: "https://github.com/testowner/testrepo.git"},
+		// 					{Key: "COMMIT_SHA", Value: "0c894fe31c2e91d59cb1a59bb25aaa78089919c2"},
+		// 				},
+		// 			}},
+		// 		},
+		// 		&bspb.BuildEvent{
+		// 			Id: &bspb.BuildEventId{Id: &bspb.BuildEventId_BuildMetadata{}},
+		// 			Payload: &bspb.BuildEvent_BuildMetadata{BuildMetadata: &bspb.BuildMetadata{
+		// 				// Status reporting is only enabled for CI builds.
+		// 				Metadata: map[string]string{"ROLE": "CI"},
+		// 			}},
+		// 		},
+		// 	},
+		// },
+	} {
+		t.Run(test.name, func(t *testing.T) {
+			te := testenv.GetTestEnv(t)
+			fakeGH := &FakeGitHubStatusService{}
+			te.SetGitHubStatusService(fakeGH)
+			auth := testauth.NewTestAuthenticator(testauth.TestUsers("USER1", "GROUP1"))
+			te.SetAuthenticator(auth)
+			ctx := context.Background()
+			handler := build_event_handler.NewBuildEventHandler(te)
+
+			// Start an invocation
+			seq := NewBESSequence(t)
+			channel := handler.OpenChannel(ctx, seq.InvocationID)
+
+			// Handle Started event referencing the metadata events as children
+			var metadataEventIDs []*bspb.BuildEventId
+			for _, e := range test.metadataEvents {
+				metadataEventIDs = append(metadataEventIDs, e.GetId())
+			}
+			started := &bspb.BuildEvent{
+				Id:       &bspb.BuildEventId{Id: &bspb.BuildEventId_Started{}},
+				Children: metadataEventIDs,
+				Payload: &bspb.BuildEvent_Started{Started: &bspb.BuildStarted{
+					Command: "build",
+					// TODO: the test fails unless OptionsDescription is set,
+					// which seems error-prone.
+					OptionsDescription: "--some_build_options",
+				}},
+			}
+			err := channel.HandleEvent(seq.NextRequest(started))
+			require.NoError(t, err)
+
+			// Should not have published any statuses yet
+			require.Empty(t, fakeGH.Clients)
+
+			// Publish all but the last metadata event - no statuses should be
+			// published.
+			md := test.metadataEvents
```

Comment:
I'm probably missing the point, but I don't understand the point of this block - this logic feels very specific and like it might makes this test difficult to extend

Reply:
Added some clarification in the comments - this is testing that the order in which we receive metadata events shouldn't matter, and that we wait to receive _all_ metadata before publishing a status to GitHub (otherwise we'd wind up publishing an inaccurate status).